common/expected.hpp

```diff
@@ -22,153 +22,24 @@
-using ExpectedString = Expected<std::string, error::Error>;
-using ExpectedBytes = expected::Expected<std::vector<uint8_t>, error::Error>;
-using ExpectedInt = Expected<int, error::Error>;
-using ExpectedLong = Expected<long, error::Error>;
-using ExpectedLongLong = Expected<long long, error::Error>;
-using ExpectedBool = Expected<bool, error::Error>;
-using ExpectedSize = Expected<size_t, error::Error>;
-
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType>::Expected(const Expected &e) {
-	if (e.has_val_) {
-		this->has_val_ = true;
-		new (&this->ex_) ExpectedType(e.value());
-	} else {
-		this->has_val_ = false;
-		new (&this->err_) ErrorType(e.error());
-	}
-}
-
-
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType>::Expected(Expected &&e) {
-	if (e.has_val_) {
-		this->has_val_ = true;
-		new (&this->ex_) ExpectedType(std::move(e.value()));
-	} else {
-		this->has_val_ = false;
-		new (&this->err_) ErrorType(e.error());
-	}
-}
-
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType>::Expected(const ExpectedType &ex) :
-	has_val_(true),
-	ex_(ex) {};
-
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType>::Expected(ExpectedType &&ex) :
-	has_val_(true),
-	ex_(std::move(ex)) {};
-
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType>::Expected(const ErrorType &err) :
-	has_val_(false),
-	err_(err) {};
-
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType>::~Expected() {
-	if (this->has_val_) {
-		this->ex_.~ExpectedType();
-	} else {
-		this->err_.~ErrorType();
-	}
-};
-
-template <typename ExpectedType, typename ErrorType>
-ExpectedType &Expected<ExpectedType, ErrorType>::value() {
-	assert(this->has_val_);
-	return this->ex_;
-};
-
-template <typename ExpectedType, typename ErrorType>
-const ExpectedType &Expected<ExpectedType, ErrorType>::value() const {
-	assert(this->has_val_);
-	return this->ex_;
-};
-
-template <typename ExpectedType, typename ErrorType>
-ErrorType Expected<ExpectedType, ErrorType>::error() const {
-	assert(!this->has_val_);
-	return this->err_;
-};
-
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType> &Expected<ExpectedType, ErrorType>::operator=(const Expected &e) {
-	if (this->has_val_) {
-		this->ex_.~ExpectedType();
-	} else {
-		this->err_.~ErrorType();
-	}
-
-	this->has_val_ = e.has_val_;
-	if (e.has_val_) {
-		this->has_val_ = true;
-		new (&this->ex_) ExpectedType(e.value());
-	} else {
-		this->has_val_ = false;
-		new (&this->err_) ErrorType(e.error());
-	}
-	return *this;
-}
+using namespace std;
 
-template <typename ExpectedType, typename ErrorType>
-Expected<ExpectedType, ErrorType> &Expected<ExpectedType, ErrorType>::operator=(Expected &&e) {
-	if (this->has_val_) {
-		this->ex_.~ExpectedType();
-	} else {
-		this->err_.~ErrorType();
-	}
+using namespace tl;
```

Comment:
I think this will break in the future, because both `std` and `tl` will contain `expected` then. A better approach is maybe to remove the two `using namespace` declarations and replace with this:

```c++
#include <version>

#ifdef __cpp_lib_expected
using expected = std::expected;
#else
using expected = tl::expected;
#endif
```

See [this page](https://en.cppreference.com/w/cpp/feature_test) for more info.

Reply:
Nice! Thanks for digging this up!!